src/snowflake/snowpark/_internal/server_connection.py

```diff
@@ -171,6 +171,21 @@ def __init__(
             except TypeError:
                 pass
 
+        # thread safe param protection
+        self._thread_safe_session_enabled = self._get_client_side_session_parameter(
+            "PYTHON_SNOWPARK_ENABLE_THREAD_SAFE_SESSION", False
+        ) or self._lower_case_parameters.get(
+            "python_snowpark_enable_thread_safe_session", False
+        )
+        self._lock = (
+            threading.RLock() if self._thread_safe_session_enabled else DummyLock()
+        )
+        self._thread_store = (
+            threading.local()
+            if self._thread_safe_session_enabled
+            else DummyThreadLocal()
```

Comment:
DummyThreadLocal is just an empty class that lets us use the same interface as thread local without creating a thread local object. When we say multithreading is enabled, we are skipping creation of locks and thread-local objects.

Reply:
I guess that should be fine, i was mainly want to keep the code the same as original, where you can init the cursor on the creation of SingleThreadLocal like what it does before, instead of wait until the call of cursor property.